File: turing-backend/agent-runtime/src/agents/generalAssistant.ts
import type { ModelProviderId, ToolCallBeacon, ToolPolicyDecision } from "@turing/shared-types";
import type { AgentExecutor } from "../executor/types.js";
import type { LlmProvider } from "../llm/provider.js";
import type { RuntimeApprovalState } from "../approvals/approvalPolling.js";
import { runAuthorizedMcpTool } from "../tools/toolRunner.js";

type McpToolClient = {
  callTool(name: string, args: Record<string, unknown>, approvalToken?: string): Promise<unknown>;
};

export type GeneralAssistantTools = {
  systemMcpClient: McpToolClient;
  filesMcpClient: McpToolClient;
  postBeacon: (beacon: ToolCallBeacon) => Promise<ToolPolicyDecision>;
  getApproval: (approvalId: string) => Promise<RuntimeApprovalState>;
};

export function createGeneralAssistantExecutor(
  providers: Partial<Record<ModelProviderId, LlmProvider>>,
  tools?: GeneralAssistantTools
): AgentExecutor {
  return {
    agentId: "general_assistant",
    async *execute(job, context) {
      const trimmed = job.payload.userText.trim();

      // Debug slash commands that exercise the MCP/approval pipeline end-to-end
      // without needing LLM function calling. Kept simple on purpose for v1.0.
      if (tools && trimmed === "/tool system.time") {
        try {
          const result = await runAuthorizedMcpTool({
            agentId: "general_assistant",
            runId: job.runId,
            traceId: job.traceId,
            serverName: "system",
            toolName: "system.time",
            args: {},
            mcpClient: tools.systemMcpClient,
            postBeacon: tools.postBeacon,
            getApproval: tools.getApproval
          });
          const content = JSON.stringify(result);
          yield {
            type: "event",
            event: { sessionId: job.sessionId, runId: job.runId, traceId: job.traceId, type: "message.delta", payload: { messageId: job.assistantMessageId, delta: content } }
          };
          yield { type: "complete", content };
        } catch (error) {
          yield { type: "fail", code: "tool_call_failed", message: error instanceof Error ? error.message : "Tool call failed", retryable: false };
        }
        return;
      }

      if (tools && trimmed === "/tool files.create") {
        const args = { path: "runtime-smoke.txt", content: "created through approval flow" };
        try {
          const result = await runAuthorizedMcpTool({
            agentId: "general_assistant",
            runId: job.runId,
            traceId: job.traceId,
            serverName: "files",
            toolName: "files.create",
            args,
            mcpClient: tools.filesMcpClient,
            postBeacon: tools.postBeacon,
            getApproval: tools.getApproval
          });
          const content = JSON.stringify(result);
          yield {
            type: "event",
            event: { sessionId: job.sessionId, runId: job.runId, traceId: job.traceId, type: "message.delta", payload: { messageId: job.assistantMessageId, delta: content } }
          };
          yield { type: "complete", content };
        } catch (error) {
          yield { type: "fail", code: "tool_call_failed", message: error instanceof Error ? error.message : "Tool call failed", retryable: false };
        }
        return;
      }

      const provider = providers[job.modelProvider];
      if (!provider) {
        yield { type: "fail", code: "model_provider_unavailable", message: `Provider ${job.modelProvider} is not configured`, retryable: false };
        return;
      }

      let content = "";
      yield {
        type: "event",
        event: { sessionId: job.sessionId, runId: job.runId, traceId: job.traceId, type: "message.started", payload: { messageId: job.assistantMessageId, role: "assistant" } }
      };

      for await (const event of provider.streamChat({ model: job.model, messages: context.messages.concat([{ role: "user", content: job.payload.userText }]) })) {
        if (event.type === "delta") {
          content += event.text;
          yield {
            type: "event",
            event: { sessionId: job.sessionId, runId: job.runId, traceId: job.traceId, type: "message.delta", payload: { messageId: job.assistantMessageId, delta: event.text } }
          };
        }
        if (event.type === "error") {
          yield { type: "fail", code: event.code, message: event.message, retryable: false };
          return;
        }
      }

      yield {
        type: "event",
        event: { sessionId: job.sessionId, runId: job.runId, traceId: job.traceId, type: "message.completed", payload: { messageId: job.assistantMessageId, content } }
      };
      yield { type: "complete", content };
    }
  };
}

File: turing-backend/agent-runtime-go/internal/agent/general_assistant.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	turingv1 "github.com/mcasillas17/TuringAgent/gen/turing/v1/go/turing/v1"
	"github.com/mcasillas17/TuringAgent/turing-backend/agent-runtime-go/internal/llm"
	"github.com/mcasillas17/TuringAgent/turing-backend/agent-runtime-go/internal/safejson"
	"github.com/mcasillas17/TuringAgent/turing-backend/agent-runtime-go/internal/tools"
)

type MessageClient interface {
	FetchMessages(ctx context.Context, sessionID string) ([]llm.ChatMessage, error)
}

type GeneralAssistantTools struct {
	SystemMCP tools.MCPClient
	FilesMCP  tools.MCPClient
	Runner    *tools.Runner
}

type GeneralAssistant struct {
	providers map[turingv1.ModelProvider]llm.Provider
	messages  MessageClient
	tools     *GeneralAssistantTools
}

func NewGeneralAssistant(providers map[turingv1.ModelProvider]llm.Provider, messages MessageClient, toolset *GeneralAssistantTools) *GeneralAssistant {
	return &GeneralAssistant{providers: providers, messages: messages, tools: toolset}
}

func (a *GeneralAssistant) SetToolBeaconPoster(post func(context.Context, *turingv1.ToolCallBeacon) (*turingv1.ToolPolicyDecision, error)) {
	if a.tools == nil || a.tools.Runner == nil {
		return
	}
	a.tools.Runner.PostBeacon = post
}

func (a *GeneralAssistant) Execute(ctx context.Context, job *turingv1.AgentJob, emit func(*turingv1.RuntimeUpdate) error) error {
	if job == nil {
		return fmt.Errorf("job is required")
	}
	messages, err := a.messages.FetchMessages(ctx, job.GetSessionId())
	if err != nil {
		return emitRunFailed(emit, job, "message_fetch_failed", err.Error(), false)
	}
	if err := emit(messageEvent(job, turingv1.TuringEventType_TURING_EVENT_TYPE_MESSAGE_STARTED, map[string]any{"messageId": job.GetAssistantMessageId(), "role": "assistant"})); err != nil {
		return err
	}
	trimmed := strings.TrimSpace(job.GetUserText())
	if handled, err := a.tryDebugTool(ctx, job, trimmed, emit); handled || err != nil {
		return err
	}
	provider := a.providers[job.GetModelProvider()]
	if provider == nil {
		return emitRunFailed(emit, job, "model_provider_unavailable", fmt.Sprintf("Provider %s is not configured", job.GetModelProvider().String()), false)
	}
	requestMessages := append([]llm.ChatMessage{}, messages...)
	requestMessages = append(requestMessages, llm.ChatMessage{Role: "user", Content: job.GetUserText()})
	events, err := provider.StreamChat(ctx, llm.ChatRequest{Model: job.GetModel(), Messages: requestMessages})
	if err != nil {
		return emitRunFailed(emit, job, "model_stream_failed", err.Error(), false)
	}
	content := ""
	for event := range events {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		switch event.Type {
		case "delta":
			content += event.Text
			if err := emit(messageEvent(job, turingv1.TuringEventType_TURING_EVENT_TYPE_MESSAGE_DELTA, map[string]any{"messageId": job.GetAssistantMessageId(), "delta": event.Text})); err != nil {
				return err
			}
		case "error":
			code := event.Code
			if code == "" {
				code = "model_error"
			}
			message := event.Message
			if message == "" {
				message = code
			}
			return emitRunFailed(emit, job, code, message, false)
		}
	}
	if ctx.Err() != nil {
		return ctx.Err()
	}
	if err := emit(messageEvent(job, turingv1.TuringEventType_TURING_EVENT_TYPE_MESSAGE_COMPLETED, map[string]any{"messageId": job.GetAssistantMessageId(), "content": content})); err != nil {
		return err
	}
	return emit(&turingv1.RuntimeUpdate{Update: &turingv1.RuntimeUpdate_RunCompleted{RunCompleted: &turingv1.RuntimeRunCompleted{RunId: job.GetRunId(), AssistantMessageId: job.GetAssistantMessageId(), Content: content}}})
}

func (a *GeneralAssistant) tryDebugTool(ctx context.Context, job *turingv1.AgentJob, trimmed string, emit func(*turingv1.RuntimeUpdate) error) (bool, error) {
	if a.tools == nil || a.tools.Runner == nil {
		return false, nil
	}
	var client tools.MCPClient
	serverName := ""
	toolName := ""
	args := map[string]any{}
	switch trimmed {
	case "/tool system.time":
		client = a.tools.SystemMCP
		serverName = "system"
		toolName = "system.time"
	case "/tool files.create":
		client = a.tools.FilesMCP
		serverName = "files"
		toolName = "files.create"
		args = map[string]any{"path": "runtime-smoke.txt", "content": "created through approval flow"}
	default:
		return false, nil
	}
	if client == nil {
		return true, emitRunFailed(emit, job, "tool_call_failed", "MCP client is not configured", false)
	}
	result, err := a.tools.Runner.Run(ctx, tools.RunInput{AgentID: turingv1.AgentId_AGENT_ID_GENERAL_ASSISTANT, RunID: job.GetRunId(), TraceID: job.GetTraceId(), ServerName: serverName, ToolName: toolName, Args: args, MCPClient: client})
	if err != nil {
		return true, emitRunFailed(emit, job, "tool_call_failed", err.Error(), false)
	}
	data, err := json.Marshal(result)
	if err != nil {
		return true, emitRunFailed(emit, job, "tool_call_failed", err.Error(), false)
	}
	content := string(data)
	if err := emit(messageEvent(job, turingv1.TuringEventType_TURING_EVENT_TYPE_MESSAGE_DELTA, map[string]any{"messageId": job.GetAssistantMessageId(), "delta": content})); err != nil {
		return true, err
	}
	if err := emit(messageEvent(job, turingv1.TuringEventType_TURING_EVENT_TYPE_MESSAGE_COMPLETED, map[string]any{"messageId": job.GetAssistantMessageId(), "content": content})); err != nil {
		return true, err
	}
	return true, emit(&turingv1.RuntimeUpdate{Update: &turingv1.RuntimeUpdate_RunCompleted{RunCompleted: &turingv1.RuntimeRunCompleted{RunId: job.GetRunId(), AssistantMessageId: job.GetAssistantMessageId(), Content: content}}})
}

func messageEvent(job *turingv1.AgentJob, eventType turingv1.TuringEventType, payload map[string]any) *turingv1.RuntimeUpdate {
	structPayload, err := safejson.ToStruct(payload)
	if err != nil {
		structPayload = nil
	}
	return &turingv1.RuntimeUpdate{Update: &turingv1.RuntimeUpdate_Event{Event: &turingv1.TuringEvent{SessionId: job.GetSessionId(), RunId: job.GetRunId(), TraceId: job.GetTraceId(), Type: eventType, Payload: structPayload}}}
}

func emitRunFailed(emit func(*turingv1.RuntimeUpdate) error, job *turingv1.AgentJob, code string, message string, retryable bool) error {
	return emit(&turingv1.RuntimeUpdate{Update: &turingv1.RuntimeUpdate_RunFailed{RunFailed: &turingv1.RuntimeRunFailed{RunId: job.GetRunId(), Code: code, Message: message, Retryable: retryable}}})
}
